### ISSUE TITLE
Build failure under MXE for target x86_64 (64 bit Windows)

### ISSUE THREAD

**Issue by Caysho** [tora-tool/tora#15]
The build gets over half way, but fails with the following:

/home/toradev/tora/src/core/utils.cpp: In static member function 'static HINSTANCE__* Utils::toLibrary::loadLibrary(const QFileInfo&)':
/home/toradev/tora/src/core/utils.cpp:1100:74: warning: passing NULL to non-pointer argument 3 of 'HINSTANCE__* LoadLibraryExA(LPCSTR, HANDLE, DWORD)' [-Wconversion-null]
        handle = LoadLibraryEx( libpath.toStdString().c_str(), NULL, NULL);
^
/home/toradev/tora/src/core/utils.cpp: In static member function 'static void* Utils::toLibrary::lookupSymbol(HINSTANCE__* const&, const char*)':
/home/toradev/tora/src/core/utils.cpp:1229:31: error: invalid conversion from 'FARPROC {aka long long int (*)()}' to 'void*' [-fpermissive]
        return GetProcAddress (handle, symbol);

Looking at the code, it appears to be due to some #defines, specifically Q_OS_WIN32 vs Q_OS_WIN64

E.g. For this specific error, [1] has the following:

    void * toLibrary::lookupSymbol(LHandle const& handle, char const* symbol)
    {
#ifndef Q_OS_WIN32
        void *retval = dlsym(handle, symbol);
        const char *dlsym_error = dlerror();
        if ( !retval || dlsym_error)
        {
            // TODO throw here?
            return NULL;
        }
        return retval;
#else
        return GetProcAddress (handle, symbol);
#endif
    }

So for Q_OS_WIN64, I think the wrong code path is executed.
Changing it to the following appears to work:

    void * toLibrary::lookupSymbol(LHandle const& handle, char const* symbol)
    {
#ifndef Q_OS_WIN32
#ifndef Q_OS_WIN64
        void *retval = dlsym(handle, symbol);
        const char *dlsym_error = dlerror();
        if ( !retval || dlsym_error)
        {
            // TODO throw here?
            return NULL;
        }
        return retval;
#else
        return GetProcAddress (handle, symbol);
#endif
#endif
    }

However, it then fails after building tora.exe:
Linking CXX executable tora.exe
[100%] Built target tora
make[2]: *** No rule to make target 'src/libtora.dll.a', needed by 'src/libporacle.dll'.  Stop.
CMakeFiles/Makefile2:310: recipe for target 'src/CMakeFiles/poracle.dir/all' failed
make[1]: *** [src/CMakeFiles/poracle.dir/all] Error 2
Makefile:137: recipe for target 'all' failed
make: *** [all] Error 2

[1] https://github.com/tora-tool/tora/blob/master/src/core/utils.cpp

**Comment by Caysho** [comment]
Forgot to include build environment details:

Installed MXE in a VM, into /usr/src/mxe
Set target as x86_64-w64-mingw32.shared
Ran:
make boost
make db
make qt5
make qscintilla2

Put Oracle Instant Client into ORACLE_HOME

export MXE_DIR=/usr/src/mxe
export MXE_TARGET=x86_64-w64-mingw32.shared
export ORACLE_HOME=$MXE_DIR/usr/$MXE_TARGET/instantclient_12_1

git clone https://github.com/tora-tool/tora.git ~/tora3

Then:
toradev@toradev:~/tora/DBUILD$ $MXE_DIR/usr/bin/$MXE_TARGET-cmake     -DBOOST_ROOT=$MXE_DIR/usr/$MXE_TARGET/include/     -DQT5_BUILD=ON     -DENABLE_PGSQL=ON     -DENABLE_DB2=OFF     -DENABLE_TERADATA=OFF     -DWANT_INTERNAL_LOKI=ON     -DWANT_INTERNAL_QSCINTILLA=OFF    -DORACLE_OCI_VERSION=12C -DCMAKE_BUILD_TYPE=Release -DWIN32=ON ..
-- Could NOT find Subversion (missing:  Subversion_SVN_EXECUTABLE) 
-- Boost version: 1.57.0
-- Found Boost libraries for debug build: /usr/src/mxe/usr/x86_64-w64-mingw32.shared/include
-- TOra 3.0.0alpha will be built to install into /usr/src/mxe/usr/x86_64-w64-mingw32.shared
-- 
-- LIB_SUFFIX variable is not defined. It will be autodetected now.
-- You can set it manually with -DLIB_SUFFIX=<value> (64 for example)
-- CMAKE_SIZEOF_VOID_P=8
--  LIB_SUFFIX is set to '64'
-- 
-- non-debug build
-- Using internal loki: /home/toradev/tora/extlibs/loki/include
Qt Found OK
CMake Warning at cmake/modules/FindQScintilla.cmake:85 (MESSAGE):
  LIBPATH
  "/usr/src/mxe/usr/x86_64-w64-mingw32.shared/qt5/bin;/usr/lib/;/usr/local/lib"
Call Stack (most recent call first):
  CMakeLists.txt:232 (FIND_PACKAGE)


-- Found QScintilla2: /usr/src/mxe/usr/x86_64-w64-mingw32.shared/qt5/lib/libqscintilla2.a
--          includes: /usr/src/mxe/usr/x86_64-w64-mingw32.shared/qt5/include/Qsci
QScintilla2 Found OK
-- ORACLE_LIBRARY_CLNTSH /usr/src/mxe/usr/x86_64-w64-mingw32.shared/instantclient_12_1/oci.dll
-- ORACLE_LIBRARY_OCCI 
-- ORACLE_LIBRARY_XML 
-- ORACLE_LIBRARY_OCIEI 
-- Found Oracle: /usr/src/mxe/usr/x86_64-w64-mingw32.shared/instantclient_12_1/oci.dll (ORACLE_HOME='/usr/src/mxe/usr/x86_64-w64-mingw32.shared/instantclient_12_1')
--     includes: /usr/src/mxe/usr/x86_64-w64-mingw32.shared/instantclient_12_1/sdk/include
-- Found XML Oracle:  
Oracle OCI Library Found OK
-- PostgreSQL environment found: /usr/src/mxe/usr/x86_64-w64-mingw32.shared/include /usr/src/mxe/usr/x86_64-w64-mingw32.shared/bin/libpq.dll
-- DB2 support is disabled by user choice
-- Teradata support disabled (enable with -DENABLE_TERADATA=on)
-- Oracle OCI Library Found OK
-- -- No Oracle XDK found: ignoring
-- poracle /usr/src/mxe/usr/x86_64-w64-mingw32.shared/instantclient_12_1/oci.dll  tora trotl
-- Could NOT find Doxygen (missing:  DOXYGEN_EXECUTABLE) 
-- Configuring done
-- Generating done
-- Build files have been written to: /home/toradev/tora/DBUILD

Then run make.

**Comment by ibre5041** [comment]
Case 1:
According to docs: http://doc.qt.io/qt-4.8/qtglobal.html#Q_OS_WIN32 Q_OS_WIN32 should be defined on all Windows platforms. But maybe there is a problem in QTBUG-23120,  Q_OS_WIN32. Or maybe the problem is somewhere else. I think #defines are OK here. GetProcAddress Should be called for both Win32 and Win64. But maybe we need explicit cast here:  `return (void*)GetProcAddress (handle, symbol);`
Another possibility would be to remove this code at all, and use QLibrary class.

Case 2:
Windows libs are different from from Linux (.ELF) ones. If you want to create a plugin (i.e. a library which depends on main .exe file) then - at the time of linking - you must create file `tora.exp`. This file contains definitions of all symbols as if tora was a library. Then when linking the plugin poracle.dll you tell the linker that some undefined symbols come from other "library" tora. I do not know how to fix this issue.

**Comment by Caysho** [comment]
According to https://bugreports.qt.io/browse/QTBUG-23120 Q_OS_WIN should be used instead, with Q_OS_WIN32 and Q_OS_WIN64 being used when needing to identify Win32 vs Win64.

So in function toLibrary::lookupSymbol Q_OS_WIN32 becomes Q_OS_WIN.
I suspect other Q_OS_WIN32 references need to be changed to facilitate compiling to Win64.

Thanks for the tip about the differences in linking requirements.

I use Windows at work, and I found the build environment to be cumbersome, so I wanted to look at cross compiling under Linux to Windows.  MXE provides a good environment for this, and it would be possible to automate the process, including building the msi files with https://wiki.gnome.org/msitools 
This also means I can isolate the build environment from my workstation.

**Comment by ibre5041** [comment]
AFAIK this reflects Windows behaviour. _WIN32 is also defined for both 32 and 64 platforms. The bug is still open. And Q_OS_WIN is defined on Qt5 only.

**Comment by Caysho** [comment]
Case 1: Putting in the cast resolves it, thanks.

Case 2: Yes, I have discovered this is very much non-trivial; unfortunately I have no solution, but I will document my findings.

I got a better understanding of the library types used in Windows development [1].
As of now, MXE builds cmake v3.0.2.  There is a cmake wiki entry on how to build a DLL [2], the basics of which I tried as follows:

I created a src/windows/libtora.cpp file to get the libtora.dll.a generated.
It does not have all the required symbols, I just wanted to see if I could get the library generated.
It is a copy of the dummy.cpp file with the following includes:
resource.h and 
../core/tora_export.h

Changes to the CMakeLists.txt file as per [2]:

  IF(MINGW)
    include (GenerateExportHeader)
    add_library(tora.dll STATIC src/windows/libtora.cpp)
    GENERATE_EXPORT_HEADER(tora.dll
                           BASE_NAME libtora
                           EXPORT_MACRO_NAME libtora_EXPORT
                           EXPORT_FILE_NAME libtora_Export.h
                           STATIC_DEFINE libtora_BUILT_AS_STATIC
                           )
  ENDIF()

When the change is made to tora/CMakeLists.txt the following files are created:
DBUILD/libtora_Export.h and
DBUILD/libtora.dll.a file
However, it is at the wrong folder level.
When the code is put into tora/src/CMakeLists.txt only the libtora_Export.h file is created.  I have no idea why.

In addition to this, I gather there is actually more coding involved as per the discussion at [3].
It seems that these difficulties have to some extent been addressed by Kitware at [4] in cmake 3.4 which has command line option to do DLL symbol export.

[1] http://binglongx.com/2008/10/01/dll-lib-def-and-exp-file-types/
[2] https://cmake.org/Wiki/BuildingWinDLL
[3] http://stackoverflow.com/questions/24872225/a-simple-example-of-using-cmake-to-build-a-windows-dll
[4] http://www.kitware.com/blog/home/post/939
